<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Привет еще раз. Спасибо, что доделал работу. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Привет, Дмитрий! Спасибо, что прислал задание:) Поздравляю с первым сданным проектом. Меня зовут Слепцов Артем и я буду проверять твой проект) Ты проделал большую работу над проектом, однако его еще можно улучшить. Будет здорово, если ты, надеюсь, не против, если я буду на ты, будешь отвечать на комментарии и участвовать в диалоге. 

Мои комментарии обозначены пометкой **Комментарий ревьюера**. Далее в файле ты сможешь найти их в похожих ячейках (если фон комментария <font color=green>зелёный</font> - всё сделано правильно. Рекомендации укажу таким же цветом; <font color=orange>оранжевый</font> - некритичные замечания. Если таких замечаний меньше трех - проект может быть принят без их отработки; <font color=red>красный</font> - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения данного проекта. 

</div>

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Правильно, что есть краткое вступление в работу, описание того, что надо делать. В работе следует приводить информацию о входных данных: какие столбцы есть в таблице, их названия и какую информацию они несут. Цель работы также хорошо бы вынести в отдельный блок вступления. Еще лучше будет, если ты сделаешь план работы с гиперссылками. Да, есть плагин ToC. Но хорошо бы уметь реализовывать интерактивный план проекта вручную. Как это реализовать - смотри по [ссылке](https://stackoverflow.com/questions/49535664/how-to-hyperlink-in-a-jupyter-notebook/49717704). Так работа выглядит презентабельно. Подробнее смотри методичку по оформлению проектов. Она располагается в Notion в разделе "Как оформлять проект".

</div>


<h1><center><font size="5"><b>Содержание</b></font></center></h1>



<h1><center><font size="3">
    
[1. Открытие данных](#1)  
  
    
[2. Предобработка данных](#2)      
[2.1  Обработка пропусков](#2.1)  
[2.2  Замена типа данных](#2.2)  
[2.3  Обработка дубликатов](#2.3)  
[2.4  Лемматизация](#2.4)  
[2.5  Категоризация данных](#2.5)  
    
    
[3. Ответы на вопросы](#3)  
[•	Есть ли зависимость между наличием детей и возвратом кредита в срок?](#3.1)  
[•	Есть ли зависимость между семейным положением и возвратом кредита в срок?](#3.2)  
[•	Есть ли зависимость между уровнем дохода и возвратом кредита в срок?](#3.3)  
[•	Как разные цели кредита влияют на его возврат в срок?](#3.4)  

 
[4. Общий вывод](#4)  

</font></center></h1>



<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Отлично! Стоит также описать те данные, с которыми мы работаем.  

</div>

<a id='1'></a>
## Открытие данных

In [1]:
import pandas as pd
data=pd.read_csv('/datasets/data.csv')
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [2]:
data.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

Просмотрим общую информацию по таблице:

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Интересные значения в количестве детей (`children`) минимальное -1, возможно это опечатка, но стоит изучить подробнее. Максимальное количество детей - 20 выглядит также не очень правдоподобно, хотя возможно. Учитывая, что количество детей является важным фактором в исследовании, то нужно будет проверить это значение. 
В количестве рабочих дней (`days_employed`) наверняка есть ошибка, которая была видна еще в начале. У нас тут и отрицательные и положительные значения. Среднее значение в ~63тыс. дней, что составляет ~172,6 года не может соответствовать действительности.  Минимальное значение возраста (`dob_years`) равно 0, вполне возможно сведения не указаны. Не мог же банк выдать кредит новорожденному?

Проверим, выдавал ли банк кредит людям, у которых значение `dob_years` и `days_employed` будут нулевыми. 
Возможно среди них есть те, кто никогда не работал и ничего не заработал.

In [5]:
data[(data['total_income'].isnull() == True) & (data['days_employed'].isnull() == True)].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
children            2174 non-null int64
days_employed       0 non-null float64
dob_years           2174 non-null int64
education           2174 non-null object
education_id        2174 non-null int64
family_status       2174 non-null object
family_status_id    2174 non-null int64
gender              2174 non-null object
income_type         2174 non-null object
debt                2174 non-null int64
total_income        0 non-null float64
purpose             2174 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


2174 - не маленькое значение. Нужно узнать подробнее.

In [6]:
data[(data['total_income'].isnull() == True) & (data['days_employed'].isnull() == True)]['income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

Разные люди разных профессий. Скорее всего данные просто не заполнены.  

Проверим еще несколько интересных граф:

In [7]:
unicle_family_status = len(data['family_status'].unique())
unicle_gender = len(data['gender'].unique())
print('Уникальных семейных статусов:', unicle_family_status) 
print('Уникальных гендеров:', unicle_gender) 

Уникальных семейных статусов: 5
Уникальных гендеров: 3


Семейных статусов может быть несколько, это нормально, хотя бы их не сотня, значит сведения занесены корректно и однообразно, но возможно среди них есть пропуск. А вот количество гендеров - 3 вызывает вопросы. Вполне возможно данные просто пропущены. 

Узнаем количество значений, по которым возникли вопросы:

In [8]:
print('Людей с нулевым возрастом: ', data[data['dob_years'] == 0]['dob_years'].count())
print('Людей с количеством детей меньше 0: ', data[data['children'] < 0]['children'].count())
print('Людей с количеством детей -1: ', data[data['children'] == -1]['children'].count())
print('Людей с 20 детьми: ', data[data['children'] == 20]['children'].count())
print('Людей с 10 и более детьми: ', data[data['children'] >= 10]['children'].count())
print('Людей с пропущенным полом: ', data[(data['gender'] != 'F') & (data['gender'] != 'M')]['gender'].count())

Людей с нулевым возрастом:  101
Людей с количеством детей меньше 0:  47
Людей с количеством детей -1:  47
Людей с 20 детьми:  76
Людей с 10 и более детьми:  76
Людей с пропущенным полом:  1


Людей с нулевым возрастом 101, это много. Однозначно имеем пропуск. Банк все таки выдает кредит и паспортные данные они проверяют.
Количество детей -1 и меньше нуля одинаковое. Это опечатка, нужно исправить -1 на 1.
Количество людей с 20 детьми и больше 10 одинаковое, 20 детей достаточно уникальный случай, но встречаться 76 раз в списке из 21525 значений никак не может. Это опечатка, нужно исправить 20 на 2.
Всего одна строка с пропущенным полом. В масштабах данного исследования этот пропуск не значителен. В целом можно рассмотреть возможность заменить его любым значением, данных о фамиллии и имени у нас все равно нет, для наглядности проверим какое все таки значение указано:    

In [9]:
data[(data['gender'] != 'F') & (data['gender'] != 'M')]['gender']

10701    XNA
Name: gender, dtype: object

Оказался не пропуск.

**Вывод**

В представленных данных имеется подробная информация о клиентах, которым выданы кредиты. Анализируя сводную информацию, начало и конец таблицы можно сделать предвариетльные вывод о структуре данных, с которыми предстоит работать, их значениях, количестве и объеме. Однозначно требуется предобработка данных для структурирования информации, например: исправление регистра, приведение целей кредита в графе `purpose` к общим единым наименованиям. Дополнительно требуется выяснить, почему в стоблце таблицы имеются отрицательные значения вместе с положительными. По проведенному предварительныму анализу можно сделать выводы о наличии пропусков и не корректных данных в отдельных строках, которые необходимо исправить.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Первый взгляд на таблицу выполнен. Здорово, что ты так подробно рассмотрел имеющиеся данные и нашел в них проблемы. Давай приступать к анализу полученной информации. 

</div>

<a id='2'></a>
## Предобработка данных

<a id='2.1'></a>
### Обработка пропусков


####  Пропуски доходов

Найдем пропуски в графе доходов (`total_income`) отсортируя таблицу по типам занятости:

In [10]:
data.loc[data['total_income'].isnull(), 'income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

Определим среднее и медианное значение доходов для понимания какое значение для замены подойдет лучше:

In [11]:
mean_total_income = data.groupby('income_type')['total_income'].mean()
mean_total_income

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        170898.309923
компаньон          202417.461462
пенсионер          137127.465690
предприниматель    499163.144947
сотрудник          161380.260488
студент             98201.625314
Name: total_income, dtype: float64

In [12]:
median_total_income = data.groupby('income_type')['total_income'].median()
median_total_income

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

По некоторым типам занятости отличий нет вовсе, либо не значительны. Заменим проущенные данные на медианное значение.

In [13]:
data.loc[(data['total_income'].isnull()) & (data['income_type'] == 'сотрудник'), 'total_income'] = data.loc[data['income_type'] =='сотрудник', 'total_income'].median()
data.loc[(data['total_income'].isnull()) & (data['income_type'] == 'компаньон'), 'total_income'] = data.loc[data['income_type'] =='компаньон', 'total_income'].median()
data.loc[(data['total_income'].isnull()) & (data['income_type'] == 'пенсионер'), 'total_income'] = data.loc[data['income_type'] =='пенсионер', 'total_income'].median()
data.loc[(data['total_income'].isnull()) & (data['income_type'] == 'госслужащий'), 'total_income'] = data.loc[data['income_type'] =='госслужащий', 'total_income'].median()
data.loc[(data['total_income'].isnull()) & (data['income_type'] == 'предприниматель'), 'total_income'] = data.loc[data['income_type'] =='предприниматель', 'total_income'].median()

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Для заполнения пропусков лучше использовать метод fillna. В остальном все верно, тип занятости сильно влияет на получаемый доход. Здорово, что ты использовал данное наблюдение при обработке пропусков. 
</div>

Проверим:

In [14]:
data.loc[data['total_income'].isnull(), 'income_type'].value_counts()

Series([], Name: income_type, dtype: int64)

Пропуски устранены.

#### Пропуски возраста

Исправим пропуски в значениях возраста (`dob_years`). Определим в каких местах имеются пропуски отсортировав по типам занятости:

In [15]:
data[data['dob_years'] <= 0]['income_type'].value_counts()   #при значениях возраста от 0 до 18 результат одинаковый

сотрудник      55
компаньон      20
пенсионер      20
госслужащий     6
Name: income_type, dtype: int64

Исправим пропуски на медианные значения по каждому типу занятости:

In [16]:
data.loc[(data['dob_years'] <= 0) & (data['income_type'] == 'сотрудник'), 'dob_years'] = data.loc[data['income_type'] =='сотрудник', 'dob_years'].median()
data.loc[(data['dob_years'] <= 0) & (data['income_type'] == 'пенсионер'), 'dob_years'] = data.loc[data['income_type'] =='пенсионер', 'dob_years'].median()
data.loc[(data['dob_years'] <= 0) & (data['income_type'] == 'компаньон'), 'dob_years'] = data.loc[data['income_type'] =='компаньон', 'dob_years'].median()
data.loc[(data['dob_years'] <= 0) & (data['income_type'] == 'госслужащий'), 'dob_years'] = data.loc[data['income_type'] =='госслужащий', 'dob_years'].median()

Проверим:

In [17]:
data[data['dob_years'] <= 0]['income_type'].value_counts() 

Series([], Name: income_type, dtype: int64)

Пропуски устранены.

#### Пропуски трудового стажа

В процессе изучения таблицы в первом разделе были выявлены пустые значения и определено количество. Выявлено, что `dob_years` и `days_employed` пропущены одновременно. Поменяем на медианные значения с учетом возраста и типа занятости. 

Сформируем возврастные группы людей на основе 25%, 50%, 75% и отдельно выделим группу "пенсионеры" с возрастом более 60 лет.

In [18]:
data['dob_years'].describe()

count    21525.000000
mean        43.496167
std         12.231538
min         19.000000
25%         34.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

Создадим функцию, которая разделит на возрастные группы

In [19]:
def age_group(age):
    if age <= 34:
        return 'молодой возраст'
    if age <= 43:
        return 'зрелый возраст'
    if age <= 53:
        return 'средний возраст'
    if age < 60:
        return 'пожилой возраст'
    else: 
        return 'пенсионеры'     
    
data['age_group'] = data['dob_years'].apply(age_group)
data.head(5)


children  days_employed  dob_years education  education_id  \
0         1   -8437.673028       42.0    высшее             0   
1         1   -4024.803754       36.0   среднее             1   
2         0   -5623.422610       33.0   Среднее             1   
3         3   -4124.747207       32.0   среднее             1   
4         0  340266.072047       53.0   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose        age_group  
0               покупка жилья   зрелый возраст  
1     приобретение автомобиля   зрелый возраст  
2               покупка жилья  молодой возраст  
3  дополнительное образование  молодой возраст  
4             сыграть свадьбу  средний возраст

Определим медианные значения по выделенным возрастным группам:

In [20]:
median_days_employed = data.groupby(age_group)['days_employed'].median()
median_days_employed

зрелый возраст     -840.475201
молодой возраст   -1701.008832
пенсионеры        -1202.570219
пожилой возраст   -2424.809749
средний возраст   -2475.924789
Name: days_employed, dtype: float64

Заменим пропуски на медианные значения по выделенным группам. 

In [21]:
data.loc[(data['days_employed'].isnull()) & (data['age_group'] == 'молодой возраст'), 'days_employed'] = data.loc[data['age_group'] =='молодой возраст', 'days_employed'].median()
data.loc[(data['days_employed'].isnull()) & (data['age_group'] == 'зрелый возраст'), 'days_employed'] = data.loc[data['age_group'] =='зрелый возраст', 'days_employed'].median()
data.loc[(data['days_employed'].isnull()) & (data['age_group'] == 'средний возраст'), 'days_employed'] = data.loc[data['age_group'] =='средний возраст', 'days_employed'].median()
data.loc[(data['days_employed'].isnull()) & (data['age_group'] == 'пожилой возраст'), 'days_employed'] = data.loc[data['age_group'] =='пожилой возраст', 'days_employed'].median()
data.loc[(data['days_employed'].isnull()) & (data['age_group'] == 'пенсионер'), 'days_employed'] = data.loc[data['age_group'] =='пенсионер', 'days_employed'].median()

Проверяем исправления:

In [22]:
data[(data['total_income'].isnull() == True) & (data['days_employed'].isnull() == True)].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 13 columns):
children            0 non-null int64
days_employed       0 non-null float64
dob_years           0 non-null float64
education           0 non-null object
education_id        0 non-null int64
family_status       0 non-null object
family_status_id    0 non-null int64
gender              0 non-null object
income_type         0 non-null object
debt                0 non-null int64
total_income        0 non-null float64
purpose             0 non-null object
age_group           0 non-null object
dtypes: float64(3), int64(4), object(6)
memory usage: 0.0+ bytes


Пропуски исправлены.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

С параметром стажа мы не работаем. Поэтому и обрабатывать пропуски в нем необязательно. Тем более в нем мы видим большое число ошибок, которые не позволяют восстановить пропуски надежно. 
    
</div>

Далее выясним, почему значения трудового стажа, как уже было замечено в самом начале, указаны и положительными и отрицательными и среднее количество соответствует 173м годам.

Анализируя начало и конец таблицы, которые выведены на просмотр в самом начале видна закономерность, что все отрицательные значения `days_employed` соответствуют любым значениям `income_type`  кроме "пенсионер", у них данное значение является очень большим.
Проверим корректность значения `days_employed` первой позиции в таблице, взяв значение по модулю и поделим на количество дней в году:

In [23]:
8437.673028	/ 365

23.11691240547945

23 года, при учете возраста клиента в 42 года, наличие такого стажа нормально.
Аналогичная ситуация с последующими строками.
В дальнейшем можем просто брать значение по модулю. 

Проверим у кого в таблице положительные и отрицательные количества стажа:

In [24]:
data.loc[data['days_employed'] > 0, 'income_type'].value_counts()

пенсионер      3443
безработный       2
Name: income_type, dtype: int64

In [25]:
data.loc[data['days_employed'] < 0, 'income_type'].value_counts()

сотрудник          11075
компаньон           5073
госслужащий         1452
пенсионер            211
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64

В итоге имеем, что в таблице у всех, кроме пенсионеров и двух безработных значения стажа имеют отрицательное значение. Выведем всю информацию по двум безработным для анализа:

In [26]:
data.loc[data['income_type'] == 'безработный']

children  days_employed  dob_years education  education_id  \
3133          1  337524.466835       31.0   среднее             1   
14798         0  395302.838654       45.0    Высшее             0   

          family_status  family_status_id gender  income_type  debt  \
3133    женат / замужем                 0      M  безработный     1   
14798  гражданский брак                 1      F  безработный     0   

        total_income                  purpose        age_group  
3133    59956.991984  покупка жилья для сдачи  молодой возраст  
14798  202722.511368             ремонт жилью  средний возраст

Количество `days_employed` сопоставимо с пенсионерами - сотни тысяч дней. Учитывать их будем вместе с пенсионерами.

**Вывод**

В данном разделе были исправлены пропущенные значения в разделах доход, возраст, трудовой стаж. 
В качестве пропущенных значений были использованы медианные значения.
Выявлено, что в трудовом стаже отцательные значения не являются как таковыми отрицательными, а записаны через знак "-". У людей с типом занятости "пенсионер" и "безработный" указаны не возможные значения трудового стажа. Вероятнее всего тут имеется системная ошибка при вводе или конвертации данных. Допускается также, что данные были занесены в какой-то другой системе исчисления, но идентифицровать её не представляется возможным.
Учитывая сложившуюся ситуацию с трудовым стажем, если в дальнейшем потребуется использовать данные о нем, то необходимо будет ввести отдельные категории для каждого типа.

#### Не верные значения количества детей

Внесем исправления:

In [27]:
data.loc[data['children'] == -1, 'children'] = 1
data.loc[data['children'] == 20, 'children'] = 2

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Ошибки в таблице исправлены. Данные подготовлены к дальнейшему анализу. 

</div>

<a id='2.2'></a>
### Замена типа данных

Заменим тип на  `int` в графах `total_income`

In [28]:
data['total_income'] = data['total_income'].astype('int')

**Вывод**

Данные приведены к целочисленным значениям. Сведения о копейках нам ни к чему.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Исходя из общей информации о таблице мы видим, что два столбца имеют вещественный тип данных, который и надо заменить на целочисленный. Эта замена произведена только для дохода. Стаж также стоит сделать целочисленным показателем. 
    
Хорошо бы освоить метод to_numeric() для будущих работ. 

</div>

<a id='2.3'></a>
### Обработка дубликатов

#### Образование

Исправим регистр на нижний в графе образование. 

In [29]:
data['education'] = data['education'].str.lower()
data['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

#### Дубликаты

Определим количество дубликатов:

In [30]:
data.duplicated().sum()

71

Необходимо определить какого рода дубликаты:

In [31]:
data[data.duplicated(keep=False)].sort_values(by='dob_years').head(8)

children  days_employed  dob_years education  education_id  \
20297         1     -1172.1065       23.0   среднее             1   
8853          1     -1172.1065       23.0   среднее             1   
15892         0     -1172.1065       23.0   среднее             1   
19321         0     -1172.1065       23.0   среднее             1   
3452          0     -1172.1065       29.0    высшее             0   
18328         0     -1172.1065       29.0    высшее             0   
4216          0     -1172.1065       30.0   среднее             1   
6312          0     -1172.1065       30.0   среднее             1   

               family_status  family_status_id gender income_type  debt  \
20297       гражданский брак                 1      F   сотрудник     0   
8853        гражданский брак                 1      F   сотрудник     0   
15892  Не женат / не замужем                 4      F   сотрудник     0   
19321  Не женат / не замужем                 4      F   сотрудник     0   
3452         женат / замужем                 0      M   сотрудник     0   
18328        женат / замужем                 0      M   сотрудник     0   
4216         женат / замужем                 0      M   сотрудник     0   
6312         женат / замужем                 0      M   сотрудник     0   

       total_income                           purpose        age_group  
20297        142594                   сыграть свадьбу  молодой возраст  
8853         142594                   сыграть свадьбу  молодой возраст  
15892        142594  сделка с подержанным автомобилем  молодой возраст  
19321        142594  сделка с подержанным автомобилем  молодой возраст  
3452         142594        покупка жилой недвижимости  молодой возраст  
18328        142594        покупка жилой недвижимости  молодой возраст  
4216         142594  строительство жилой недвижимости  молодой возраст  
6312         142594  строительство жилой недвижимости  молодой возраст

Удаляем дубликаты:

In [32]:
data = data.drop_duplicates().reset_index(drop = True)
data.duplicated().sum()

0

**Вывод**

Данные об образовании приведены к нижнему регистру. Вполне вероятно, что данные заносятся вручную и по этой причине написаны с разным регистром. Удалены полностью дублирующиеся строки с данными. Как они попали в выгрузку и при этом расположены в разных местах таблицы, а не например подрят, что могло бы быть обоснованно случайным задвоением по техническим причинам - не понятно.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Дубликаты вызывают смещение финальных результатов, поэтому их и надо удалять. Конечно, в данных нет уникального идентификатора, однако надо уметь удалять дубликаты. Метод drop_duplicates() использован верно.

</div>

<a id='2.4'></a>
### Лемматизация

Произведем проверку в графе цель кредита (`purpose`), т.к. сами цели однотипны, но написаны по-разному. Их пожно обобщить по ключевым словам.

In [33]:
data['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

In [34]:
from pymystem3 import Mystem
m = Mystem()
def lemmatize(i):
    lemma = m.lemmatize(i)
    return lemma

data['new_lemmas'] = data['purpose'].apply(lemmatize)        
data.sample(5)

children  days_employed  dob_years education  education_id  \
16895         0    -649.824464       56.0   среднее             1   
1223          0   -6855.901021       43.0   среднее             1   
16450         1   -2311.174000       49.0   среднее             1   
6180          1   -5236.451085       38.0    высшее             0   
14356         2   -1883.592462       35.0    высшее             0   

          family_status  family_status_id gender income_type  debt  \
16895   женат / замужем                 0      M   сотрудник     1   
1223    женат / замужем                 0      F   сотрудник     0   
16450   женат / замужем                 0      F   сотрудник     0   
6180   гражданский брак                 1      F   сотрудник     0   
14356   женат / замужем                 0      F   сотрудник     0   

       total_income                     purpose        age_group  \
16895        254796        сделка с автомобилем  пожилой возраст   
1223         177699  строительство недвижимости   зрелый возраст   
16450        115013        сделка с автомобилем  средний возраст   
6180         207534               покупка жилья   зрелый возраст   
14356        198037  покупка жилой недвижимости   зрелый возраст   

                                     new_lemmas  
16895         [сделка,  , с,  , автомобиль, \n]  
1223       [строительство,  , недвижимость, \n]  
16450         [сделка,  , с,  , автомобиль, \n]  
6180                    [покупка,  , жилье, \n]  
14356  [покупка,  , жилой,  , недвижимость, \n]

In [35]:
from collections import Counter
all_unique_word = Counter(data['new_lemmas'].sum())
sorted(all_unique_word.items(), key=lambda pair: pair[1], reverse=True)

[(' ', 33570),
 ('\n', 21454),
 ('недвижимость', 6351),
 ('покупка', 5897),
 ('жилье', 4460),
 ('автомобиль', 4306),
 ('образование', 4013),
 ('с', 2918),
 ('операция', 2604),
 ('свадьба', 2324),
 ('свой', 2230),
 ('на', 2222),
 ('строительство', 1878),
 ('высокий', 1374),
 ('получение', 1314),
 ('коммерческий', 1311),
 ('для', 1289),
 ('жилой', 1230),
 ('сделка', 941),
 ('дополнительный', 906),
 ('заниматься', 904),
 ('проведение', 768),
 ('сыграть', 765),
 ('сдача', 651),
 ('семья', 638),
 ('собственный', 635),
 ('со', 627),
 ('ремонт', 607),
 ('подержанный', 486),
 ('подержать', 478),
 ('приобретение', 461),
 ('профильный', 436)]

**Вывод**

Произведена лемматизация целей кредита. Из представленных значений можно выделить цели: жилье, автомобиль, образование, свадьба, недвижимость.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Круто, ты сделал самую сложную часть проекта. Использовал нужные методы, и даже выделил основные цели кредитов 👍 

</div>

<a id='2.5'></a>
### Категоризация данных

#### Категоризация по доходу

Произведем категорицаю данных по показателю доходов людей. Hазделение произведем по квантилям, так мы получим группы почти одинакого размера. Возьмем разделение с шагом в 25%, которое мы уже видели в первом разделе исследования при изучении исходных данных.

In [36]:
pd.qcut(data['total_income'], q=4).value_counts(sort=False)

(20666.999, 107623.0]     5364
(107623.0, 142594.0]      5479
(142594.0, 195820.25]     5247
(195820.25, 2265604.0]    5364
Name: total_income, dtype: int64

Применим функцию для разделения на категории по выделенным квантилям:

In [37]:
group_1 = 107623
group_2 = 142594
group_3 = 195820
group_4 = 2265604

def total_income_cut(total_income):
    if total_income <= group_1:
        return 'Группа 1'
    if total_income <= group_2:
        return 'Группа 2'
    if total_income <= group_3:
        return 'Группа 3'
    else:
        return 'Группа 4'
    
data['total_income_cut'] = data['total_income'].apply(total_income_cut)
data.sample(10)

children  days_employed  dob_years            education  education_id  \
2907          0    -273.781036       58.0               высшее             0   
13207         0    -188.363973       50.0               высшее             0   
2716          0   -1516.375426       23.0              среднее             1   
3064          0  -12392.309374       56.0              среднее             1   
15006         2   -1831.696426       35.0  неоконченное высшее             2   
7149          0   -4046.868469       52.0              среднее             1   
1644          0  342113.056132       67.0              среднее             1   
10491         2   -2202.429854       33.0              среднее             1   
3923          2    -426.152484       33.0              среднее             1   
18676         1   -1411.331777       33.0              среднее             1   

               family_status  family_status_id gender income_type  debt  \
2907         женат / замужем                 0      F   компаньон     0   
13207  Не женат / не замужем                 4      F   компаньон     0   
2716        гражданский брак                 1      M   сотрудник     0   
3064         женат / замужем                 0      M   сотрудник     1   
15006        женат / замужем                 0      M   сотрудник     0   
7149               в разводе                 3      M   сотрудник     0   
1644         женат / замужем                 0      F   пенсионер     0   
10491        женат / замужем                 0      F   сотрудник     0   
3923         женат / замужем                 0      M   сотрудник     0   
18676        женат / замужем                 0      F   сотрудник     0   

       total_income                                 purpose        age_group  \
2907         172357  строительство собственной недвижимости  пожилой возраст   
13207        201879   получение дополнительного образования  средний возраст   
2716          95025                                 свадьба  молодой возраст   
3064         128418           получение высшего образования  пожилой возраст   
15006        145390           получение высшего образования   зрелый возраст   
7149         111697              дополнительное образование  средний возраст   
1644          66803         операции со своей недвижимостью       пенсионеры   
10491        196568                   заняться образованием  молодой возраст   
3923         102370       покупка коммерческой недвижимости  молодой возраст   
18676         75970                              автомобиль  молодой возраст   

                                              new_lemmas total_income_cut  
2907   [строительство,  , собственный,  , недвижимост...         Группа 3  
13207  [получение,  , дополнительный,  , образование,...         Группа 4  
2716                                       [свадьба, \n]         Группа 1  
3064         [получение,  , высокий,  , образование, \n]         Группа 2  
15006        [получение,  , высокий,  , образование, \n]         Группа 3  
7149                [дополнительный,  , образование, \n]         Группа 2  
1644     [операция,  , со,  , свой,  , недвижимость, \n]         Группа 1  
10491                   [заниматься,  , образование, \n]         Группа 4  
3923     [покупка,  , коммерческий,  , недвижимость, \n]         Группа 1  
18676                                   [автомобиль, \n]         Группа 1

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Исходя из чего выбран именно такой способ категоризации столбца с доходами? Группы получились достаточными по размеру, но все же выбор способа стоит обосновать подробнее. 
</div>
<div class="alert alert-success">
    
Лучше использовать разбиение по квантилям данных (метод qcut). Тогда группы получатся одинакового размера. Или же можно узнать в интернете процентный состав общества по доходам и затем уже категоризовать столбец с доходами по найденному соотношению с помощью персентилей. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Так мы получаем равные по размеру группы и можем надежнее провести анализ. 

</div>

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

Категоризовать клиентов по наличию детей необязательно. Разбиение есть уже в исходных данных. Все-таки разница в возврате кредита у семей с одним ребенком или двумя детьми будет. Так ты только уменьшаешь объем имеющейся информации. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Убрать лишнюю категоризацию - верное решение. 

</div>

**Вывод**

Произведены разделения данных на категории по ежемесячному доходу. Категории сформированы по раному количеству людей, в таком случае при анализе завимостей возврата кредита от дохода в равной степенеи будут анализироваться все группы людей.

#### Категоризация по цели кредита

Ранее были выделены леммы по основным целям: жилье, автомобиль, образование, свадьба, недвижимость. Произведем замену значений в таблице.

In [38]:
def purpose_category(i):
    if 'жилье' in i:
        return 'жилье'        
    if 'автомобиль' in i:
        return 'автомобиль'
    if 'образование' in i:
        return 'образование'
    if 'свадьба' in i:
        return 'свадьба'
    if 'недвижимость' in i:
        return 'недвижимость'
    else:
        return 'ОШБИКА'    # данное значение создаем для того, что бы при подсчете значений мы точно знали, 
                           # что не все категории были применены
    
data['purpose_category'] = data['new_lemmas'].apply(purpose_category)
data.sample(10)

children  days_employed  dob_years education  education_id  \
13118         0            NaN       61.0   среднее             1   
4237          1   -5147.883548       30.0   среднее             1   
20615         0   -4021.010281       47.0   среднее             1   
11214         0   -3610.581955       55.0   среднее             1   
18214         0  347051.864200       63.0   среднее             1   
15392         0   -1541.851658       46.0    высшее             0   
21411         1    -865.081118       25.0   среднее             1   
2865          1   -2521.934229       30.0   среднее             1   
10093         0  372543.317309       60.0   среднее             1   
20089         1   -1556.378102       29.0   среднее             1   

          family_status  family_status_id gender  income_type  debt  \
13118         в разводе                 3      F    пенсионер     0   
4237    женат / замужем                 0      F    сотрудник     0   
20615   женат / замужем                 0      F    сотрудник     0   
11214    вдовец / вдова                 2      F    сотрудник     0   
18214   женат / замужем                 0      F    пенсионер     0   
15392   женат / замужем                 0      F    сотрудник     0   
21411   женат / замужем                 0      F    компаньон     0   
2865   гражданский брак                 1      F  госслужащий     1   
10093   женат / замужем                 0      F    пенсионер     0   
20089  гражданский брак                 1      F    сотрудник     0   

       total_income                   purpose        age_group  \
13118        118514      сделка с автомобилем       пенсионеры   
4237          85786    профильное образование  молодой возраст   
20615        156495  операции с недвижимостью  средний возраст   
11214        115064     на покупку автомобиля  пожилой возраст   
18214        166790        высшее образование       пенсионеры   
15392        198698              ремонт жилью  средний возраст   
21411         94065           свой автомобиль  молодой возраст   
2865         276740           свой автомобиль  молодой возраст   
10093         46805     получение образования       пенсионеры   
20089         61388     на проведение свадьбы  молодой возраст   

                                  new_lemmas total_income_cut purpose_category  
13118      [сделка,  , с,  , автомобиль, \n]         Группа 2       автомобиль  
4237        [профильный,  , образование, \n]         Группа 1      образование  
20615  [операция,  , с,  , недвижимость, \n]         Группа 3     недвижимость  
11214    [на,  , покупка,  , автомобиль, \n]         Группа 2       автомобиль  
18214          [высокий,  , образование, \n]         Группа 3      образование  
15392                 [ремонт,  , жилье, \n]         Группа 4            жилье  
21411              [свой,  , автомобиль, \n]         Группа 1       автомобиль  
2865               [свой,  , автомобиль, \n]         Группа 4       автомобиль  
10093        [получение,  , образование, \n]         Группа 1      образование  
20089    [на,  , проведение,  , свадьба, \n]         Группа 1          свадьба

Определяем общее количество кредитов по категориям:

In [39]:
data['purpose_category'].value_counts() 

недвижимость    6351
жилье           4460
автомобиль      4306
образование     4013
свадьба         2324
Name: purpose_category, dtype: int64

По сводной статистике видно, что самая популярная статья кредитов это недвижимость, в целом категории жилье и недвижимость можно было бы и объединить, но не всегда невижимость может быть жилой и приобретается с целью жилья. Жилье, автомобиль, образование составляют примерно равные доли. Кредит на свадьбу берут меньше всего.

<div class="alert alert-danger">
<font size="5"><b>Комментарий ревьюера</b></font>

В данных есть словари. Попробуй их обнаружить и выделить. Это довольно распространенный способ хранения информации. Мне важно понять, что он тобой освоен. Заметь, что в основной таблице мы имеем по 2 параметра с образованием и семейным статусом. Именно по ним и стоит выделить словари. 

</div>

#### Словари по группам образования и семейного статуса

Создадим словари из имеющихся однотипных данных в графах образования и семейного статуса и соотношением по id

In [40]:
education_dict=data.set_index('education')['education_id'].to_dict()
education_dict

{'высшее': 0,
 'среднее': 1,
 'неоконченное высшее': 2,
 'начальное': 3,
 'ученая степень': 4}

In [41]:
family_status_dict=data.set_index('family_status')['family_status_id'].to_dict()
family_status_dict

{'женат / замужем': 0,
 'гражданский брак': 1,
 'вдовец / вдова': 2,
 'в разводе': 3,
 'Не женат / не замужем': 4}

Созданы словари из данных.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Верно, теперь мы можем убрать текстовые параметры из основной таблицы, что увеличит эффективность работы с данными. 

</div>

<a id='3'></a>
## Ответы на вопросы

<a id='3.1'></a>
- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Построим сводную таблицу по категориям людей с детьми и невозвратом кредита с указанием количества таких людей:

In [42]:
data_pivot = data.pivot_table(index = ['children'], values = 'debt', aggfunc=['mean', 'count'])
data_pivot.sort_values(by='children', ascending = False)

mean  count
              debt   debt
children                 
5         0.000000      9
4         0.097561     41
3         0.081818    330
2         0.094925   2128
1         0.091658   4855
0         0.075438  14091

**Вывод**

Люди, у которые нет детей имеют 7,5% невозврата кредитов, что является наименьшим значением. Количество таких людей больше половины.
Люди с детьми имеют более высокий процент невозврата кредита. Исключением в данном анализе составлет только люди с пятью детьми, но их количество сравнительно мало и этим может быть обусловлено нулевое значение. В любом случае наличие детей сказывается на фактах просрочек выплат.

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Посмотри на результаты по полным данным без группировки. 

</div>

<a id='3.2'></a>
- Есть ли зависимость между семейным положением и возвратом кредита в срок?

Построим сводную таблицу:

In [43]:
data_pivot = data.pivot_table(index = ['family_status'], values = 'debt', aggfunc=['mean', 'count'])
data_pivot.sort_values(by=('mean','debt'), ascending = True)

mean  count
                           debt   debt
family_status                         
вдовец / вдова         0.065693    959
в разводе              0.071130   1195
женат / замужем        0.075452  12339
гражданский брак       0.093471   4151
Не женат / не замужем  0.097509   2810

**Вывод**

Люди, которые состояли в браке реже невозвращают кредит, чем те, которые в нем состоят или не состоят вовсе.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Вывод не противоречит полученному результату. 

</div>

<a id='3.3'></a>
- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Построим сводную таблицу:

In [44]:
data_pivot = data.pivot_table(index = ['total_income_cut'], values = 'debt')
data_pivot.sort_values(by=('total_income_cut'), ascending = True)

debt
total_income_cut          
Группа 1          0.079605
Группа 2          0.088155
Группа 3          0.085382
Группа 4          0.071402

**Вывод**

Проверка зависимости производится в равных по количеству людей группах. Люди с самым малым и самым большим доходом меньше всех склонны к невозвратам кредитов. Люди со средним доходом более склонны к невозвратам кредитов. Возможно стоит предположить, что люди с малыми и приблюжающимися к средним доходами более ограничены финансово, что влияет на расчет и сознательность. В противоположность им люди из категории самых богатых не имеют явных финансовых ограничений, которые есть в первой группе и которые могли бы способствовать просрочке выплат. К тому же более богатые люди очевидно берут кредиты на более высокие суммы и выплата штрафа за просрочку в натуральном выражении составляет не малую сумму.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Подумай над тем, почему мы получаем такой результат.

</div>

<a id='3.4'></a>
- Как разные цели кредита влияют на его возврат в срок?

Построим сводную таблицу:

In [45]:
data_pivot = data.pivot_table(index = ['purpose_category'], values = 'debt', aggfunc=['mean', 'count'])
data_pivot.sort_values(by=('mean', 'debt'), ascending= True)

mean count
                      debt  debt
purpose_category                
жилье             0.069058  4460
недвижимость      0.074634  6351
свадьба           0.080034  2324
образование       0.092200  4013
автомобиль        0.093590  4306

**Вывод**

Наименьший невозврат кредитов в срок с целью покупки жилья и недвижимости. Кредиты на автотранспорт и образование имеют практически одинаковые значения невозврата.

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

Здорово, что при анализе использован метод сводных таблиц. Он пригодится в дальнейших проектах.

Выводы сделаны по всем пунктам. 

</div>

<a id='4'></a>
## Общий вывод

Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

Построим сводную таблицу, которая отражает семейное положение и количество детей с указанием количества таких заемщиков:

In [46]:
data_pivot = data.pivot_table(index = ['family_status','children'], values = 'debt', aggfunc=['mean', 'count'])
data_pivot.sort_values(by='family_status', ascending = False)

mean count
                                    debt  debt
family_status         children                
женат / замужем       5         0.000000     7
                      4         0.103448    29
                      3         0.068273   249
                      2         0.093552  1582
                      1         0.082224  3004
                      0         0.069095  7468
гражданский брак      5         0.000000     2
                      4         0.000000     8
                      3         0.142857    56
                      2         0.092958   355
                      1         0.118000  1000
                      0         0.083883  2730
вдовец / вдова        4         0.000000     1
                      3         0.000000     6
                      2         0.125000    24
                      1         0.086420    81
                      0         0.062574   847
в разводе             4         0.000000     1
                      3         0.090909    11
                      2         0.096386    83
                      1         0.066456   316
                      0         0.070153   784
Не женат / не замужем 1         0.114537   454
                      4         0.500000     2
                      3         0.125000     8
                      2         0.119048    84
                      0         0.092838  2262

Безусловно зависимость наличия детей и отсутствия узаконенных отношений негавтино влияют на факт погашения кредита в срок. Люди с детьми в статусе не женат / не замужем или в гражданском браке чаще всего не могут погасить кредит в срок. Идеальный заемщик отличается наличием либо малого, либо большого дохода. Наличие брака в текущий или прошлый момент и отсутствие детей положительно сказываются на выплате кредита в срок.

<div class="alert alert-warning">
<font size="5"><b>Комментарий ревьюера</b></font>

Выводы описаны грамотно. Приведены ответы на главные вопросы проекта. В выводах можно приводить полученные ранее значения, верно. Также можно расписать все, что было сделано в ходе проведения работы. Вот только блок с описанием самого плохо заемщика рекомендую убрать. Пересечение признаков мы не смотрим, да и посмотри на число клиентов в группах выше. Это 1 или 2 клиента. Делать выводы по такому числу может быть рискованно. 

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера</b></font>

- В начале работы следует описывать данные, с которыми работаешь;
- Не проводи лишние категоризации;
- Выдели словари в данных.

Ты проделал большую работу, молодец! Ты овладел важными на данном этапе обучения инструментами, а также довольно качественно их используешь. Критических замечаний немного. Однако их важно исправить. Также есть достаточное число желтых комментариев. Думаю, исправить и их тебе не составит труда. Жду твою работу :)

</div>

<div class="alert alert-success">
<font size="5"><b>Комментарий ревьюера 2</b></font>

Помарки исправлены, и теперь работа выполнена хорошо. У тебя получился классный проект, молодец. Продолжай в том же духе! Успехов в дальнейших работах :)

</div>
   

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.